information/specification.md

```diff
@@ -0,0 +1,706 @@
+complicated than it would initially appear.
+
+ * Each transaction has a _maximum size_, as defined by the protocol. The size
+   of a transaction increases as we add more inputs or outputs.
+
+   Therefore, there's a practical limit on the number of coins we can select
+   for any given transaction.
+
+ * The most obvious strategy for coin selection, which consists of trying to
+   get as close to the requested value as possible, will tend (over time) to
+   create a lot of _dust_: small unspent outputs.
+
+   Dust outputs are a problem, because even if the total value of dust in a
+   wallet is more than enough to cover a given target amount, if we attempt to
+   include that dust in a given transaction, we may run out of space (by
+   reaching the transaction size limit) before we can cover the target amount.
+
+ * The most obvious strategy for paying change, which consists of making a
+   single change output with the exact excess value, will tend (over time) to
+   reduce the size of the UTxO set. This is bad for two reasons:
+
+    1. Having a small UTxO set limits the number of payments that we can make
+       in parallel. Since a single UTxO entry can only be used to pay for a
+       single output, we need at least as many UTxO entries as there are
+       outputs.
+
+    2. The approach of coalescing all change into a single output is widely
+       considered to have negative privacy implications. The discussion of
+       privacy issues is beyond the scope of this article.
+
+### Goals
+
+There are several desirable properties that we would like a coin selection
+algorithm to have. These properties include:
+
+ * A coin selection algorithm should, over the course of time, aim to generate
+   and maintain a UTxO set with _useful_ outputs: that is, outputs that allow
+   us to process future payments with a _minimum_ number of inputs.
+
+ * A coin selection algorithm should employ strategies to limit the
+   amount of _dust_ that accumulates in the UTxO set.
+
+For more information on dust avoidance, see [Self Organisation in Coin
+Selection](#self-organization-in-coin-selection).
+
+# Definitions
+
+### Address
+
+An _address_ is a unique identifier that represents a destination for a payment.
+
+Addresses are owned (and generated) by individual wallets.
+
+In general, coin selection algorithms are agnostic to the choice of address
+format. Therefore, the details of individual address formats are not described
+within this document.
+
+### Coin Amount
+
+A _coin amount_ is a non-negative integer value that represents a number of
+[Lovelace](https://cardanodocs.com/cardano/monetary-policy/).
+
+One [Ada](https://cardanodocs.com/cardano/monetary-policy/) is _exactly_ equal
+to one million Lovelace.
+
+### UTxO Set
+
+A _UTxO set_ represents the unspent value associated with a wallet.
+
+Each member of the set is a pair of the form (**_i_**, **_o_**), where:
+
+  * **_i_** is a [transaction input](#transaction-input);
```

Comment:
A UTxO is actually a transaction identifier and an output index. I am not sure to agree with the defnition here :)

Reply:
(see also: https://github.com/input-output-hk/cardano-wallet/wiki/About-UTxO )

Reply:
@KtorZ wrote:
> A UTxO is actually a transaction identifier and an output index. I am not sure to agree with the definition here

Fair enough!

When I originally wrote this article, I was attempting to be consistent with the types defined in our code, which _effectively_ defines the UTxO set as a set of pairs of (`TxIn`, `TxOut`):

```hs
newtype UTxO = UTxO { getUTxO :: Map TxIn TxOut }
```

However, as far as coin selection algorithms are concerned, the `UTxO` could also be defined as:

```hs
newtype UTxO k = UTxO { getUTxO :: Map k Coin }
```

From the point of view of a coin selection algorithm, so long as `k` is instantiated to a type that provides an `Ord` instance, the other details are irrelevant. (See https://github.com/input-output-hk/cardano-coin-selection/pull/29.)

The question then becomes:
> How can we relate this type to the human-readable definition that we provide?

We could write something like this:

> A _UTxO set_, or the set of _unspent transaction outputs_, represents the unspent value associated with a wallet.
>
> From the point of view of a coin selection algorithm, each entry in the UTxO set consists of a pair of the form (**_o_**, **_v_**), where:
>   * **_o_** is a [transaction output](#transaction-output);
>   * **_v_** is a [coin value](#coin-value).

Which can then reuse our _existing_ definition of _transaction output_:
> a pair (**_h_**, **_n_**), where:
>  * **_h_** is a _unique identifier_ for an existing transaction _t_;
 > * **_n_** is a 0-based integer index into the output list of transaction _t_.

Thoughts?

Reply:
@KtorZ I've now completely revised this section with new definitions.

See: https://github.com/input-output-hk/cardano-coin-selection/pull/16/commits/4f266317a99ca3ccf87f1c98a302bb3f617f2a25